packages/react/src/components/Table/TableHead/FilterHeaderRow/FilterHeaderRow.jsx

```diff
@@ -373,11 +378,37 @@ class FilterHeaderRow extends Component {
           const isLastColumn = visibleColumns.length - 1 === i;
           const lastVisibleColumn = visibleColumns.slice(-1)[0];
           const isLastVisibleColumn = column.id === lastVisibleColumn.columnId;
+          /* istanbul ignore next */
+          const CustomInput = column.customInput;
           // undefined check has the effect of making isFilterable default to true
           // if unspecified
           const headerContent =
             column.isFilterable !== undefined && !column.isFilterable ? (
               <div />
+            ) : /* istanbul ignore next */
+            column.customInput !== undefined ? (
+              <CustomInput
+                ref={this.setFirstFilterableRef}
+                onChange={(event) => {
+                  if (event.persist) {
+                    event.persist();
+                  }
+                  this.setState(
+                    (state) => ({
+                      filterValues: {
+                        ...state.filterValues,
+                        [column.id]: event.selectedItem
+                          ? getFilterValue(event.selectedItem)
+                          : event.selectedItems?.length > 0
+                          ? event.selectedItems.map(getMultiselectFilterValue)
+                          : event.target.value,
+                      },
+                    }),
+                    debounce(this.handleApplyFilter, 1000)
+                  );
+                }}
+                value={filterValues[column.id]}
```

Comment:
We should pass as much information to the CustomInput so that someing that is rendering the input can have as much information about the data.  We should send columns and column so that the input component can make other decisions around what to render (ie, it can use the information in the column as well)